dev-util/serve-d/serve-d-0.8.0_beta15-r2.ebuild

```diff
@@ -0,0 +1,213 @@
+		# Only look for them in the ${T} directory
+		--registry=file://"${T}"
+		# Prefer verbose operation
+		--verbose
+		# Unpack dependencies for the local project (put them in ${S}/.dub)
+		--cache=local
+	)
+	for dep in ${A}; do
+		if [[ ${dep} != ${P}.zip ]]; then
+			# Due to a bug in dub, we can't have similar looking
+			# archives in the same directory. Simply put, if we had:
+			# foo-1.zip and foo-bar-2.zip when trying to fetch foo dub
+			# will fail because it thinks that foo-bar-2.zip is foo's
+			# archive. See: https://github.com/dlang/dub/pull/2727
+			cp "${DISTDIR}/${dep}" "${T}" || die "Could not copy dependency to temporary directory"
+
+			# This calculation is faulty, dub allows digits in package names, it will
+			# probably be needed to save dependencies in a similar to $CRATES variable.
+			name="${dep%%-[0-9]*}"
+			version="${dep%.zip}"
+			version="${version#${name}-}"
+
+			"${EPREFIX}"/usr/bin/dub "${dub_args[@]}" fetch "${name}@${version}" \
+				|| die "Could not extract dub dependencies"
+
+			rm "${T}/${dep}" || die "Could not remove depedency from temporary directory"
+		fi
+	done
+	popd > /dev/null || die
+}
+
+src_compile() {
+	local dub_args=(
+		# It wouldn't be good if the build became interactive
+		--non-interactive
+		# Let user provided $DCFLAGS dictate the build
+		#--build=realease
+		# It is good practice to be verbose
+		--verbose
+		# Don't touch anything outside the ${S} directory
+		--skip-registry=all
+	)
+
+	# A little overkill but it doesn't ignore $LDFLAGS.
+	#
+	# Note, dub adds -Wl,-no-as-needed so having it be replaced by the
+	# -Wl,--as-needed common in $LDFLAGS may cause issues.
+	DFLAGS="$(get_dflags) ${DLANG_LDFLAGS}" DUB_HOME="${S}" \
+		  "${EPREFIX}"/usr/bin/dub build "${dub_args[@]}" || die
+}
+
+src_test() {
+	local dub_args=(
+		# It is good practice to be verbose
+		--verbose
+		# Don't touch anything outside the ${S} directory
+		--skip-registry=all
+
+		# ${PV} uses silly for testing so it supports jobs
+		--
+		--threads="$(makeopts_jobs)"
+	)
+
+	# Tests all submodules in dub.json. :dcd currently fails tests due
+	# to a missing dependency and import. It's 1 little test though so
+	# no biggie. See: https://github.com/Pure-D/serve-d/pull/350
+	local subpkg
+	for subpkg in ":http" ":protocol" ":lsp" ":serverbase" ":workspace-d" ""; do
+		local cmd=(
+			env
+			DUB_HOME="${S}"
+			DFLAGS="$(get_dflags) ${DLANG_LDFLAGS}"
+			"${EPREFIX}"/usr/bin/dub
+			test
+			# Let $subpkg be shell expanded since it can be empty
+			${subpkg}
+
+			# We also have another issue that testing $PV gives linking errors
+			# when the --build configuration is specified (or if $DFLAGS are specified).
+			# We can work around it by forcing --build=unittest.
+			# See: https://github.com/Pure-D/serve-d/issues/351
+			--build=unittest
+
+			"${dub_args[@]}"
+		)
+		echo "${cmd[@]}"
+		"${cmd[@]}" || die "Tests failed"
+	done
+}
+
+src_install() {
+	dobin "${S}/serve-d"
+}
+
+get_dflags() {
```

Comment:
I suppose you would want to implement `filter-dcflags` which, as an easy implementation, would take the first argument to be (dmd|gdc|ldc2) and the second one the flag to filter for that compiler, but otherwise act the same as filter-flags.

The more generic implementation would take in a impl glob as the first argument and to the same thing as above. This is a little harder as I skipped adding support for such matching of implementations because I didn't consider it useful enough.

However, the more I spend thinking about it, the more I believe that going the implementation matching path would be better, as that would allow `dlang_gen_cond_dep` which, in turn, would allow net-misc/onedrive to depend on a library that wraps curl when compiled with gdc to get this annoying edge case out of the way for good. I do see this, as well as your requested change, to be good enough arguments for implementing the matching in dlang-utils.eclass, but, just to confirm, you agree with me going all the way with the pattern-based implementation as that would take me a little bit of time to make sure that I understand the code and I would have to change `dlang_gen_dep` to `dlang_gen_cond_dep` as well.

Reply:
I've went up and added `dlang-filter-dflags` and `dlang_gen_cond_dep` like promised, it wasn't as long as I've believed.